src/supplemental/nanolib/ringbuffer/ringbuffer.c

```diff
@@ -200,52 +201,171 @@ static inline int ringBuffer_rule_check(ringBuffer_t *rb, void *data, int flag)
-		ringBuffer_clean_msgs(rb, 1);
-		return -1;
+	nng_msg **smsgs = (nng_msg **)nng_aio_get_prov_data(file->aio);
+	if (smsgs == NULL) {
+		log_error("parquet callback smsgs is NULL\n");
+		return;
 	}
 
+	parquet_file_ranges *file_ranges = (parquet_file_ranges *)nng_aio_get_output(file->aio, 1);
+	if (file_ranges == NULL) {
+		log_error("parquet file range is NULL\n");
+		return;
+	}
+
+	uint32_t *szp = (uint32_t *)nng_aio_get_msg(file->aio);
+	if (szp == NULL) {
+		log_error("parquet file size is NULL\n");
+		return;
+	}
+
+	for (int i = 0; i < file_ranges->size; i++) {
+		parquet_file_range **file_range = file_ranges->range;
+
+		ringBufferFileRange_t *range = nng_alloc(sizeof(ringBufferFileRange_t));
+		if (range == NULL) {
+			log_error("alloc new file range failed! no memory! msg will be freed\n");
+			return;
+		}
+
+		range->startidx = file_range[i]->start_idx;
+		range->endidx = file_range[i]->end_idx;
+		range->filename = nng_alloc(sizeof(char) * strlen(file_range[i]->filename));
+		if (range->filename == NULL) {
+			log_error("alloc new file range filename failed! no memory! msg will be freed\n");
+			nng_free(range, sizeof(ringBufferFileRange_t));
+			range = NULL;
+			return;
+		}
+
+		(void)strncpy(range->filename, file_range[i]->filename, strlen(file_range[i]->filename));
+
+		cvector_push_back(file->ranges, range);
+		log_warn("ringbus: parquet write to file: %s success\n", file_range[i]->filename);
+	}
+	for (uint32_t i = 0; i < *szp; i++) {
+		if (smsgs[i] != NULL) {
+			nng_msg_free(smsgs[i]);
+			smsgs[i] = NULL;
+		}
+	}
+
+	if (smsgs != NULL) {
+		nng_free(smsgs, sizeof(nng_msg *) * *szp);
+		smsgs = NULL;
+	}
+
+	return;
+}
+
+static parquet_object *init_parquet_object(ringBuffer_t *rb, ringBufferFile_t *file)
+{
+	if (rb == NULL || file == NULL) {
+		log_error("parquet object or ringbuffer is NULL\n");
+		return NULL;
+	}
+
+	uint8_t **darray = nng_alloc(sizeof(uint8_t *) * rb->size);
+	uint32_t *dsize = nng_alloc(sizeof(uint32_t) * rb->size);
+	uint64_t *keys = nng_alloc(sizeof(uint64_t) * rb->size);
+
+	if (keys == NULL || darray == NULL || dsize == NULL) {
+		log_error("alloc new keys darray dsize failed! no memory! msg will be freed\n");
+
+		if (keys != NULL) {
+			nng_free(keys, sizeof(uint64_t) * rb->size);
+		}
+		if (darray != NULL) {
+			nng_free(darray, sizeof(uint8_t *) * rb->size);
+		}
+		if (dsize != NULL) {
+			nng_free(dsize, sizeof(uint32_t) * rb->size);
+		}
+
+		return NULL;
+	}
+
+	nng_msg **smsgs = nng_alloc(sizeof(nng_msg *) * rb->size);
+	for (unsigned int i = 0; i < rb->size; i++) {
+		keys[i] = rb->msgs[i].key;
+		smsgs[i] = rb->msgs[i].data;
+		darray[i] = nng_msg_payload_ptr((nng_msg *)rb->msgs[i].data);
+		dsize[i] = nng_msg_len((nng_msg *)rb->msgs[i].data) -
+		        (nng_msg_payload_ptr((nng_msg *)rb->msgs[i].data) -
+				(uint8_t *)nng_msg_body((nng_msg *)rb->msgs[i].data));
+	}
+
+	nng_aio *aio;
+	nng_aio_alloc(&aio, ringbuffer_parquet_cb, file);
```

Comment:
so where to free?

Reply:
That's a question to think about, and I'll think about it more tomorrow